Marlin/src/module/ft_motion.cpp

```diff
@@ -541,258 +433,261 @@ void FTMotion::setTrajectoryType(const TrajectoryType type) {
     case TrajectoryType::POLY5:       currentGenerator = &poly5Generator; break;
     case TrajectoryType::POLY6:       currentGenerator = &poly6Generator; break;
   }
-  currentGenerator->reset(); // Reset the selected generator
 }
 
 // Load / convert block data from planner to fixed-time control variables.
 // Called from FTMotion::loop() at the fetch of the next planner block.
-void FTMotion::loadBlockData(block_t * const current_block) {
-  // Cache the extruder index for this block
-  TERN_(DISTINCT_E_FACTORS, block_extruder_axis = E_AXIS_N(current_block->extruder));
+// Return whether a plan is available.
+bool FTMotion::plan_next_block() {
+  while (true) {
+
+    const bool had_block = !!stepper.current_block;
+    discard_planner_block_protected();                                  // Always clears stepper.current_block...
+    block_t * const current_block = planner.get_current_block();  // ...so get the current block from the queue
+
+    // The planner had a block and there was not another one?
+    const bool planner_finished = had_block && !current_block;
+    if (planner_finished) {
+      runoutBlock();
+      return true;
+    }
```

Comment:
```suggestion
    discard_planner_block_protected();  // Always clears stepper.current_block...
    block_t * const current_block = planner.get_current_block(); // ...so get the current block from the queue

    if (!current_block) {
      if (stepper.current_block) { // The planner had a block and there was not another one?
        runoutBlock();
        return true;
      }
      currentGenerator->planRunout(0); //// There was never a block? Run out the plan and bail.
      return false;
    }
```
And remove tines 455 to 459 in ft_motion.cpp from this PR. I think that is clearer, and avoids 2 intermediate variables.

Reply:
I added the intermediate variables to be clearer (had a block and doesn't now is clearer for me), and more importantly to avoid any bugs that may come from planner.get_current_block assuming the old block shouldn't be taken into account anymore.

Reply:
The bug it fixed was releated to having to clear the block from the stepper before adding the runout block i think. Keeping the "had_block" as a boolean allows clearing the current block from the stepper earlier without affecting the logic.

Reply:
Compiler will optimize intermediate variables out, so you shouldn't avoid them if it will hurt code readability.

Reply:
It's not hot path code anyway. All of this happens once per block so it doesn't affect perf unless we add 1000s of unnecessary checks.